### PR TITLE
 Math-Magicians (Project)

### PR DESCRIPTION
**Pull request by rahimialijan**
# Math-Magicians Project (Testing the Logics)

- Installed the jest for my project.
- Create a calculate.test.js file to test my Calculate function inside the calculate.js file
- Create an operate.test.js file to test my operate function inside the operate.js file.

### CODE REVIEW

File: src/components/Calculator/logic/operate.test.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const operate = require('./operate').default;
+
+test('Addition operation should return the correct result', () => {
+  const result = operate('5', '3', '+');
+  expect(result).toBe('8');
+});
+test('Division operation should return the correct result', () => {
+  const result = operate('10', '2', '÷');
+  expect(result).toBe('5');
+});
```
Comment (Munsa1) [inline src/components/Calculator/logic/operate.test.js]: - [ ] ll the functionalities & scenarios of the operate.js should be covered using Jest. Kindly ensure that the following scenarios are tested as well.
- Multiplication of two numbers should return the correct value.
image

- Subtraction of two numbers should return the correct value.
image

- A Number cannot be divided by o
image

- A test for the `%` Modulus operation
image

- Kindly do not edit the `operate.js` file as we have to test for a case where we cannot find the `%` modulus of a number by `0`. Your `operate.js` file is missing this logic, kindly add it and test for this.

Your Operate.js
image


Given Operate.js
image

We need to test that `you can't find modulo as can't divide by 0`.
image


- A test for an unknown operation should also be covered.
image

File: src/components/Calculator/logic/calculate.test.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  const expectedState = {
+    total: null,
+    text: null,
+    operation: null,
+  };
+
+  const newState = calculate(initialState, 'AC');
+
+  expect(newState).toEqual(expectedState);
+});
```
Comment (Munsa1) [inline src/components/Calculator/logic/calculate.test.js]: - [ ] Good job adding a calculate.js test, however, we have to test all scenarios of the `calculate.test.js`. We have also been advised to use React Testing Library to simulate user interaction for these tests.
Kindly ensure you add tests for the scenarios in calculate.js including:
- What happens when a user clicks the `.` point operand
image

- what happens when the user clicks the `-/+` negative slash positive operand
image

- If there is no operation, update next and clear the value
image

- If there is an operation, update next
image

- User pressed an operation after pressing '='
image

- What happens when a user pressed an operation button and there is an existing operation
image

File: src/components/Calculator/logic/calculate.test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  const expectedState = {
+    total: null,
+    text: null,
+    operation: null,
+  };
+
+  const newState = calculate(initialState, 'AC');
+
+  expect(newState).toEqual(expectedState);
```
Comment (Munsa1) [inline src/components/Calculator/logic/calculate.test.js]: - [ ] React Testing Library should be used to simulate user interaction for this test

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @rahimialijan @supreeti, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Passing Linters :heavy_check_mark: 
Correct GitHub Flow :heavy_check_mark: 
Descriptive ReadMe File :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._
 ### Due to insufficient files changed I will leave some comments here
- [ ] Kindly remove the link to the `Live Demo` section as you have removed the section itself and this link has become broken
image

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [ ] Going forward kindly Capitalize the first letter of the first word of your commit messages

- [ ] Also you can consider adding your coding partner as a co-author to the author section of the readme file
image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Munsa1** [comment]
## Important :red_circle:
- [ ] Kindly note that you should have Jest snapshots for testing all the components in your app. These include `HomePage`, `quotes`, and the `Calculator` component itself. This is a requirement, and I have left it here due to insufficient files being changed.

**Review by AdnanOlarmmi — APPROVED** [review]
Hi @rahimialijan and @supreeti,
This is Adnan Is-haq, your beloved code reviewer, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Approved office man

## Highlights
- Handle all tests edge cases.

## Optional suggestions
_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
